### PR TITLE
Better horizontal scroll for Tree

### PR DESCRIPTION
**Pull request by timothyqiu**
* Adds pan gesture support for horizontal scroll.
* Scrolls horizontally when needed to ensure cursor visiblility.

### CODE REVIEW

**Comment by akien-mga** [comment]
For the reference: normally at this stage of development I'd be quite cautious merging any changes to core APIs like Tree, but I know that you're doing your changes with care and the resulting code for this and previous PRs seems much better.

So I'm making an exception to the release freeze by merging those, and we'll see in the upcoming 3.2 beta 5 if there are any regressions in the editor or user games; I trust that you'll be around to assess those if any come up :)

File: scene/gui/tree.cpp [hunk 1]
```diff
@@ -3455,30 +3459,48 @@ int Tree::get_item_offset(TreeItem *p_item) const {
 }
 
 void Tree::ensure_cursor_is_visible() {
-
-	if (!is_inside_tree())
+	if (!is_inside_tree()) {
 		return;
+	}
+	if (!selected_item || (selected_col == -1)) {
+		return;  // nothing under cursor
```
Comment (akien-mga) [inline scene/gui/tree.cpp]: Style issue around comments: https://travis-ci.org/godotengine/godot/jobs/632234552?utm_medium=notification&utm_source=github_status

BTW we're not enforcing it yet, but the current proposal for a comment style guide would be to start with a capital letter and end with a dot.

**Comment by timothyqiu** [comment]
Thanks for your trust. I'll be watching related issues.

**Comment by akien-mga** [comment]
Thanks!

File: scene/gui/tree.cpp [hunk 2]
```diff
@@ -3455,30 +3459,48 @@ int Tree::get_item_offset(TreeItem *p_item) const {
...
-	if (h > screenh) { //screen size is too small, maybe it was not resized yet.
-		v_scroll->set_value(ofs);
-	} else if (ofs + h > v_scroll->get_value() + screenh) {
-		v_scroll->call_deferred("set_value", ofs - screenh + h);
-	} else if (ofs < v_scroll->get_value()) {
-		v_scroll->set_value(ofs);
+		if (cell_h > screen_h) { // Screen size is too small, maybe it was not resized yet.
+			v_scroll->set_value(y_offset);
+		} else if (y_offset + cell_h > v_scroll->get_value() + screen_h) {
+			v_scroll->call_deferred("set_value", y_offset - screen_h + cell_h);
```
Comment (YeldhamDev) [inline scene/gui/tree.cpp:3482]: What's the reason for this to be deferred? It causes the edit box to be misaligned when editing an partially visible item.

**Pull request merged**